lib/rucio/transfertool/transfertool.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2018 CERN for the benefit of the ATLAS collaboration.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Martin Barisits <martin.barisits@cern.ch>, 2018
+
+from abc import ABCMeta, abstractmethod
+
+
+class Transfertool():
+    """
+    Interface definition of the Rucio transfertool
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, external_host):
+        """
+        Initializes the transfertool
+
+        :param external_host:   The external host where the transfertool API is running
+        """
+
+        self.external_host = external_host
+
+    @abstractmethod
+    def submit(self, files, job_params, timeout=None):
+        """
+        Submit transfers to the transfertool.
+
+        :param files:        List of dictionaries describing the file transfers.
+        :param job_params:   Dictionary containing key/value pairs, for all transfers.
+        :returns:            Transfertool internal identifiers.
+        """
+        pass
```

Comment:
raise NotImplementedError

Reply:
It's an abstract class, does not need the NotImplementedError as it cannot be used like this anyway.

Reply:
I'm not sure. Nothing prevent from instantiating an object from this class and submitting transfers. raise or return NotImplemented is more explicit and will fail also for the sub classes derived from that one if the method is not defined.  It's safer.

Reply:
Python does ;)
```
from rucio.transfertool.transfertool import Transfertool
class a(Transfertool):
    pass
x = a()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class a with abstract methods cancel, query, submit, update_priority
```

Reply:
good